platform/darwin/src/MGLMapSnapshotter.mm

```diff
@@ -142,20 +149,159 @@ - (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshot
 #endif
             [strongSelf drawAttributedSnapshot:attributions snapshotImage:mglImage pointForFn:pointForFn queue:queue completionHandler:completion];
         }
-        _snapshotCallback = NULL;
+        strongSelf->_snapshotCallback = NULL;
     });
 
-    dispatch_async(queue, ^{
-        __typeof__(self) strongSelf = weakSelf;
-        if (!strongSelf) {
-            return;
+    // Launches snapshot on background Thread owned by mbglMapSnapshotter
+    // _snapshotCallback->self() is an ActorRef: if the callback is destroyed, further messages
+    // to the callback are just no-ops
+    _mbglMapSnapshotter->snapshot(_snapshotCallback->self());
+}
+
++ (void)drawAttributedSnapshotWorker:(mbgl::MapSnapshotter::Attributions)attributions snapshotImage:(MGLImage *)mglImage pointForFn:(mbgl::MapSnapshotter::PointForFn)pointForFn queue:(dispatch_queue_t)queue scale:(CGFloat)scale size:(CGSize)size completionHandler:(MGLMapSnapshotCompletionHandler)completion {
+    
+    NS_ARRAY_OF(MGLAttributionInfo *)* attributionInfo = [MGLMapSnapshotter generateAttributionInfos:attributions];
```

Comment:
Just an FYI for when it comes to merge to master: `NS_ARRAY_OF`  and friends have been removed (but are still in `release-boba`)